core/build.gradle.kts

```diff
@@ -232,6 +233,81 @@ tasks {
         // maxHeapSize = "1024m"
 //        executable = "$JDK_6/bin/java"
     }
+
+    create("compileJavaModuleInfo", JavaCompile::class) {
+        val compileKotlinJvm = getByName<KotlinCompile>("compileKotlinJvm")
+        val sourceDir = file("jvm/java9/")
+        val targetFile = compileKotlinJvm.destinationDir.resolve("../module-info.class")
+
+        // Use a Java 11 compiler for the module info.
+        javaCompiler.set(project.javaToolchains.compilerFor { languageVersion.set(JavaLanguageVersion.of(11)) })
+
+        // Always compile kotlin classes before the module descriptor.
+        dependsOn(compileKotlinJvm)
+
+        // Add the module-info source file.
+        source(sourceDir)
+
+        // Also add the module-info.java source file to the Kotlin compile task.
+        // The Kotlin compiler will parse and check module dependencies,
+        // but it currently won't compile to a module-info.class file.
+        // Note that module checking only works on JDK 9+,
+        // because the JDK built-in base modules are not available in earlier versions.
+        val javaVersion = compileKotlinJvm.kotlinJavaToolchain.javaVersion.orNull
+        // Note: extra sanity check, because `jdkHome` might be overridden, this can be
+        // removed once https://github.com/Kotlin/kotlinx-datetime/pull/155/files is merged.
+        val isJava8 = compileKotlinJvm.kotlinOptions.jdkHome?.contains("1.8") == true
+        if (!isJava8 && javaVersion?.isJava9Compatible == true) {
+            logger.info("Module-info checking is enabled; $compileKotlinJvm is compiled using Java $javaVersion")
+            compileKotlinJvm.source(sourceDir)
+        } else {
+            logger.info("Module-info checking is disabled")
+        }
+
+        // This task outputs only the module-info class file.
+        outputs.file(targetFile)
+
+        // Use the same destination dir to see classes compiled by compileKotlinJvm.
+        destinationDir = compileKotlinJvm.destinationDir
```

Comment:
Let's not risk messing the output of the main compilation, better to output it in a separate dir. The output of the main compilation can be passed to JavaCompile with the `--patch-module` option